src/frontend/src/helpers/Storage/WebStorage.ts

```diff
@@ -0,0 +1,588 @@
+    super();
+    this.connection = connection;
+    // this.synced = false;
+    this.connected = false;
+    this.failed = false;
+    // this.isSyncing = false;
+    this.offlineMode = 0; // TODO change this to look up the cookie
+    this.timeoutCount = 0;
+    this.timeoutInterval = null;
+    this.key = 0;
+    this.storage = storage;
+    this.socket = new SeashellWebsocket(this.connection.wsURI, this.connection.key);
+    this.callbacks = [];
+    this.debug = debug;
+    this.socket.debug = debug;
+
+    // These functions are not available in offline mode.
+    // this.ping = this.make_offline_disabled('ping');
+    // this.newProject = this.make_offline_disabled('newProject');/
+    // this.newProjectFrom = this.make_offline_disabled('newProjectFrom');
+    // this.deleteProject = this.make_offline_disabled('deleteProject');
+    // this.restoreFileFrom = this.make_offline_disabled('restoreFileFrom');
+    // this.getUploadFileToken = this.make_offline_disabled('getUploadFileToken');
+    // this.getExportToken = this.make_offline_disabled('getExportToken');
+    // this.marmosetSubmit = this.make_offline_disabled('marmosetSubmit');
+    // These functions do nothing and just resolve in offline mode.
+    // this.lockProject = this.make_offline_noop('lockProject');
+    // this.forceLockProject = this.make_offline_noop('forceLockProject');
+    // this.unlockProject = this.make_offline_noop('unlockProject');
+    // this.archiveProjects = this.make_offline_noop('archiveProjects');
+
+    // These functions either:
+    //  - return the online result if online.
+    //  - return the offline result if offline.
+    // this.getProjects = this.make_offline_enabled('getProjects');
+    // this.getFiles = this.make_offline_enabled('getFiles');
+    // this.readFile = this.make_offline_enabled('readFile');
+    // this.getRunFile = this.make_offline_enabled('getRunFile');
+    // this.getSettings = this.make_offline_enabled('getSettings');
+    // this.getMostRecentlyUsed = this.make_offline_enabled('getMostRecentlyUsed');
+
+    // These functions:
+    //  - invoke the offline version if offline
+    //  - invoke the both the offline version and the online
+    //    version if online, returning the offline version.
+    //  - marks the local storage as dirty and queues up a sync
+    // this.newFile = this.make_offline_enabled('newFile', true);
+    // this.writeFile = this.make_offline_enabled('writeFile', true);
+    // this.deleteFile = this.make_offline_enabled('deleteFile', true);
+    // this.renameFile = this.make_offline_enabled('renameFile', true);
+    // this.setSettings = this.make_offline_enabled('setSettings', true);
+    // this.updateMostRecentlyUsed = this.make_offline_enabled('updateMostRecentlyUsed', true)
+  }
+
+  public async getRunFile(proj: ProjectID, question: string): Promise<FileID> {
+    try {
+      const result = await this.socket.sendMessage({
```

Comment:
One issue I see with that is we probably only want to have one instance of WebsocketClient, so a separate Compiler class would have to share a WebsocketClient with WebStorage. Which we could do.. but it's not as "nice".

Also it seems like converting from a tuple to a File object could be pretty neatly done in WebsocketClient, that's kind of what I imagined that layer would be for (taking raw response from server and converting to the types we work with on the frontend).

It probably is a good idea to separate the compiling/running code to a seperate interface. Let me work on that next since I was planning on hooking up the offline compiler anyway.